src/rrd/mod.rs

```diff
@@ -0,0 +1,493 @@
+                metric TEXT NOT NULL UNIQUE, 
+                value FLOAT NOT NULL,
+                PRIMARY KEY (timestamp, metric)
+                );",
+        )
+        .execute(&mut connection)
+        .await?;
+
+        sqlx::query::<Sqlite>("CREATE INDEX IF NOT EXISTS ts_index ON usage (timestamp);")
+            .execute(&mut connection)
+            .await?;
+
+        sqlx::query::<sqlx::Sqlite>("CREATE INDEX IF NOT EXISTS ts_index ON last (timestamp);")
+            .execute(&mut connection)
+            .await?;
+
+        Ok(RRDImpl {
+            pool,
+            retention: retention.as_secs() as i64,
+            window: window.as_secs() as i64,
+        })
+    }
+
+    pub async fn print<W: Write>(&mut self, mut writer: W) -> Result<W> {
+        RRDImpl::print_last_usage(self, &mut writer).await?;
+        let mut connection = self.pool.acquire().await?;
+        let timestamps: Vec<i64> = sqlx::query_scalar("SELECT DISTINCT timestamp FROM usage;")
+            .fetch_all(&mut connection)
+            .await?;
+        for ts in timestamps {
+            RRDImpl::print_ts(self, ts, &mut writer).await?;
+        }
+        Ok(writer)
+    }
+
+    async fn print_last_usage<W: Write>(&mut self, mut writer: W) -> Result<()> {
+        let mut connection = self.pool.acquire().await?;
+        writer.write_fmt(format_args!(".last\n"))?;
+        let records: Vec<(String, f64)> = sqlx::query_as("SELECT metric, usage FROM last;")
+            .fetch_all(&mut connection)
+            .await?;
+        for (metric, usage) in records {
+            writer.write_fmt(format_args!("\t{}: {}\n", metric, usage))?
+        }
+        Ok(())
+    }
+
+    async fn print_ts<W: Write>(&mut self, ts: i64, mut writer: W) -> Result<()> {
+        let mut connection = self.pool.acquire().await?;
+        let records: Vec<(String, f64)> =
+            sqlx::query_as("SELECT metric, usage FROM usage WHERE timestamp = ? ;")
+                .bind(ts)
+                .fetch_all(&mut connection)
+                .await?;
+        for (metric, usage) in records {
+            writer.write_fmt(format_args!("\t{}: {}\n", metric, usage))?
+        }
+        Ok(())
+    }
+
+    async fn retain(&mut self, now: i64) -> Result<()> {
+        // should retain be unsigned?
+        let mut connection = self.pool.acquire().await?;
+        let retain = (now - self.retention) as i64;
+        sqlx::query("DELETE FROM usage WHERE timestamp <= ? ;")
+            .bind(retain)
+            .execute(&mut connection)
+            .await?;
+        Ok(())
+    }
+
+    pub async fn slots(&mut self) -> Result<Vec<i64>> {
+        let mut connection = self.pool.acquire().await?;
+        let timestamps: Vec<i64> = sqlx::query_scalar("SELECT DISTINCT timestamp FROM usage ;")
+            .fetch_all(&mut connection)
+            .await?;
+        Ok(timestamps)
+    }
+
+    pub async fn slot_at(&mut self, ts: i64) -> Result<RRDSlotImpl> {
```

Comment:
Why slot_at is public? and why the `slot` method doesn't use this instead of re-implementing ?. I know you use it for testing as well (Same way in go) but there is no reason this becomes public.

Also it will make more sense that the `slot` method implementations actually just calls this with `now`

Reply:
You don't wanna fix bugs in the 2 places